### ISSUE TITLE
stray `#[derive(` in `possible candidate is found in another module` diagnostic

### ISSUE THREAD

**Issue by arielb1** [rust-lang/rust#44566]
## Meta

This occurs in 539f2083de809b5c8304fe7426655cfeb0e66d5e (from 13-09) and 1.21 beta, but not on stable

## STR

```Rust
#[derive(Debug)]
pub struct X;

impl Add for X {} // trigger unused import

fn main() {}
```

## Expected result

On stable, you get a correct candidate suggestion:
```
   Compiling playground v0.0.1 (file:///playground)
error[E0405]: cannot find trait `Add` in this scope
 --> src/main.rs:4:6
  |
4 | impl Add for X {} // trigger unused import
  |      ^^^ not found in this scope
  |
help: possible candidate is found in another module, you can import it into scope
  |
1 | use std::ops::Add;
  |
```

## Actual Result

The suggestion contains a stray `derive`:
```
   Compiling playground v0.0.1 (file:///playground)
error[E0405]: cannot find trait `Add` in this scope
 --> src/main.rs:4:6
  |
4 | impl Add for X {} // trigger unused import
  |      ^^^ not found in this scope
  |
help: possible candidate is found in another module, you can import it into scope
  |
1 | #[derive(use std::ops::Add; <--- STRAY DERIVE IN SUGGESTION
  |
```

cc @jseyfried @estebank

**Comment by arielb1** [comment]
Started a bisection run

**Comment by oli-obk** [comment]
dupe of #44210

**Comment by arielb1** [comment]
This is a regression to beta, and the other PR doesn't have a bisection run, so I'm continuing with it.

**Comment by oli-obk** [comment]
Also happens on nightly

**Comment by nikomatsakis** [comment]
@oli-obk do you plan to fix this and #44210?

**Comment by nikomatsakis** [comment]
triage: P-high

**Comment by nikomatsakis** [comment]
In #44210, which was closed as a duplicate, @oli-obk wrote:

> No, the first item is inside the expansion. We need a macro check. I don't think Rust's lints have any, but clippy has a convenience function in almost every lint. We should probably upstream it. There are so many diagnostics that choke on expansions.

which seems worth preserving.

**Comment by oli-obk** [comment]
I have a fix open as a PR: https://github.com/rust-lang/rust/pull/44215

It fixes the regular command line output once and for all. But the structured json suggestions are still not right. This is discussed in the PR together with possible solutions. We can merge it as it is and fix the json stuff later. No one is really expecting structured suggestions to be fully correct anyway at the moment.

**Comment by arielb1** [comment]
The bad suggestion line was moved to #44995 (because it's not a regression). Closing this issue as it is fixed and backported to beta in #44918.
